lightning/src/sign/tx_builder.rs

```diff
@@ -1,19 +1,161 @@
+#![allow(dead_code)]
 
+use core::cmp;
 use core::ops::Deref;
 
 use bitcoin::secp256k1::{self, PublicKey, Secp256k1};
 
 use crate::ln::chan_utils::{
-	commit_tx_fee_sat, htlc_success_tx_weight, htlc_timeout_tx_weight,
+	commit_tx_fee_sat, htlc_success_tx_weight, htlc_timeout_tx_weight, htlc_tx_fees_sat,
 	ChannelTransactionParameters, CommitmentTransaction, HTLCOutputInCommitment,
 };
 use crate::ln::channel::{CommitmentStats, ANCHOR_OUTPUT_VALUE_SATOSHI};
 use crate::prelude::*;
 use crate::types::features::ChannelTypeFeatures;
 use crate::util::logger::Logger;
 
+#[cfg(any(test, fuzzing))]
+#[derive(Clone, PartialEq, PartialOrd, Eq, Ord)]
+pub(crate) struct HTLCAmountDirection {
+	pub outbound: bool,
+	pub amount_msat: u64,
+}
+
+#[cfg(not(any(test, fuzzing)))]
+pub(crate) struct HTLCAmountDirection {
+	pub outbound: bool,
+	pub amount_msat: u64,
+}
+
+impl HTLCAmountDirection {
+	fn is_dust(
+		&self, local: bool, feerate_per_kw: u32, broadcaster_dust_limit_satoshis: u64,
+		channel_type: &ChannelTypeFeatures,
+	) -> bool {
+		let htlc_tx_fee_sat = if channel_type.supports_anchors_zero_fee_htlc_tx() {
+			0
+		} else {
+			let htlc_tx_weight = if self.outbound == local {
+				htlc_timeout_tx_weight(channel_type)
+			} else {
+				htlc_success_tx_weight(channel_type)
+			};
+			// As required by the spec, round down
+			feerate_per_kw as u64 * htlc_tx_weight / 1000
+		};
+		self.amount_msat / 1000 < broadcaster_dust_limit_satoshis + htlc_tx_fee_sat
+	}
+}
+
+pub(crate) struct NextCommitmentStats {
+	pub next_commitment_htlcs: Vec<HTLCAmountDirection>,
+	pub holder_balance_msat: Option<u64>,
+	pub counterparty_balance_msat: Option<u64>,
+	pub commit_tx_fee_sat: u64,
+	pub dust_exposure_msat: u64,
+	// If the counterparty sets a feerate on the channel in excess of our dust_exposure_limiting_feerate,
+	// this should be set to the dust exposure that would result from us adding an additional nondust outbound
+	// htlc on the counterparty's commitment transaction.
+	pub extra_nondust_htlc_on_counterparty_tx_dust_exposure_msat: Option<u64>,
+}
+
+fn on_holder_tx_dust_exposure_msat(
+	next_commitment_htlcs: &[HTLCAmountDirection], dust_buffer_feerate: u32,
+	broadcaster_dust_limit_satoshis: u64, channel_type: &ChannelTypeFeatures,
+) -> u64 {
+	next_commitment_htlcs
+		.iter()
+		.filter_map(|htlc| {
+			htlc.is_dust(true, dust_buffer_feerate, broadcaster_dust_limit_satoshis, channel_type)
+				.then_some(htlc.amount_msat)
+		})
+		.sum()
+}
+
+fn on_counterparty_tx_dust_exposure_msat(
+	next_commitment_htlcs: &[HTLCAmountDirection], dust_buffer_feerate: u32,
+	excess_feerate_opt: Option<u32>, broadcaster_dust_limit_satoshis: u64,
+	channel_type: &ChannelTypeFeatures,
+) -> (u64, Option<u64>) {
+	let mut on_counterparty_tx_dust_exposure_msat: u64 = next_commitment_htlcs
+		.iter()
+		.filter_map(|htlc| {
+			htlc.is_dust(false, dust_buffer_feerate, broadcaster_dust_limit_satoshis, channel_type)
+				.then_some(htlc.amount_msat)
+		})
+		.sum();
```

Comment:
Likewise here the only difference between `on_counterparty_tx_dust_exposure_msat` and `on_holder_tx_dust_exposure_msat` is the `is_local` parameter.

Perhaps we can just move this into `get_next_commitment_stats` and have a dedicated method for calculating `extra_nondust_htlc_on_counterparty_tx_dust_exposure_msat`?